File: server/viewmodels/issue.js
const StatusViewModel = require('./status')
const IssueTypeViewModel = require('./issue-type')
const EpicViewModel = require('./epic')
const CommentViewModel = require('./comment')
const UserViewModel = require('./user')
const JiraToMarkdown = require('./jira-to-markdown')
const AttachmentViewModel = require('./attachment')

module.exports = class IssueViewModel {
  constructor () {
    this.id = 0
    this.key = ''
    this.title = ''
    this.description = ''
    this.assignee = null
    this.reporter = null
    this.IssueStatus = null
    this.IssueType = null
    this.createdAt = Date.now()
    this.updatedAt = Date.now()
    this.children = []
    this.comments = []
    this.attachments = []
    this.tags = []
  }

  static createFromJira (obj, colorObj) {
    const result = new IssueViewModel()
    if (obj.fields.attachment) {
      for (let attachment of obj.fields.attachment) {
        result.attachments.push(AttachmentViewModel.createFromJira(attachment))
      }
    }
    result.id = obj.id
    result.key = obj.key
    result.title = obj.fields.summary
    result.description = JiraToMarkdown.convert(obj.fields.description, result.attachments)
    result.assignee = UserViewModel.createFromJira(obj.fields.assignee)
    result.reporter = UserViewModel.createFromJira(obj.fields.reporter)
    result.updatedAt = obj.fields.updated
    result.createdAt = obj.fields.created
    result.IssueStatus = StatusViewModel.createFromJira(obj.fields.status)
    result.issueType = IssueTypeViewModel.createFromJira(obj.fields.issuetype, colorObj)
    result.epic = EpicViewModel.createFromJira(obj.fields.epic)
    result.tags = obj.fields.labels ? obj.fields.labels : []
    const comments = obj.fields.comment ? obj.fields.comment.comments : []
    for (let comment of comments) {
      result.comments.push(CommentViewModel.createFromJira(comment))
    }
    return result
  }

  static createFromLocal (obj) {
    const result = new IssueViewModel()
    result.id = obj.id
    result.key = `AB-${obj.id}`
    result.title = obj.title
    result.description = obj.description
    result.updatedAt = obj.updatedAt
    result.createdAt = obj.createdAt
    result.assignee = UserViewModel.createFromLocal(obj.assignee)
    result.reporter = UserViewModel.createFromLocal(obj.reporter)
    result.IssueStatus = StatusViewModel.createFromLocal(obj.IssueStatus)
    result.issueType = IssueTypeViewModel.createFromLocal(obj.IssueType)
    for (let comment of obj.comments) {
      result.comments.push(CommentViewModel.createFromLocal(comment))
    }
    if (obj.children) {
      for (let child of obj.children) {
        result.children.push(IssueViewModel.createFromLocal(child))
      }
    }
    return result
  }
}

File: server/model/issue.py
from ._base import BaseModel
from .issue_status import Status
from .issue_type import IssueType
from .user import User


class Issue(BaseModel):

    def __init__(self, id_=0, description='', title='', key='', created='', updated=''):
        super().__init__(id_)
        self.description = description
        self.title = title
        self.key = key
        self.createdAt = created
        self.updatedAt = updated
        self.status = None
        self.type = None
        self.assignee = None
        self.reporter = None
        self.editor = None

    @classmethod
    def from_db_dict(cls, obj, comments=[], children=[]):
        issue = cls(
            obj.get('id'),
            obj.get('description'),
            obj.get('title'),
            obj.get('key'),
            obj.get('createdAt'),
            obj.get('updatedAt'),
        )
        status = Status(obj.get('status_id'), obj.get('status_name'))
        issue.status = status
        type_ = IssueType(
            obj.get('type_id'),
            obj.get('type_name'),
            obj.get('type_colour'),
            obj.get('type_subtask'),
        )
        issue.type = type_
        assignee = User(
            obj.get('assignee_id'),
            obj.get('assignee_externalId'),
            obj.get('assignee_username'),
            obj.get('assignee_displayName'),
            obj.get('assignee_avatar'),
            obj.get('assignee_createdAt'),
            obj.get('assignee_updatedAt'),
        )
        issue.assignee = assignee if assignee.id else None
        reporter = User(
            obj.get('reporter_id'),
            obj.get('reporter_externalId'),
            obj.get('reporter_username'),
            obj.get('reporter_displayName'),
            obj.get('reporter_avatar'),
            obj.get('reporter_createdAt'),
            obj.get('reporter_updatedAt'),
        )
        issue.reporter = reporter if reporter.id else None
        editor = User(
            obj.get('editor_id'),
            obj.get('editor_externalId'),
            obj.get('editor_username'),
            obj.get('editor_displayName'),
            obj.get('editor_avatar'),
            obj.get('editor_createdAt'),
            obj.get('editor_updatedAt'),
        )
        issue.editor = editor if editor.id else None
        return issue

    def to_viewmodel(self):
        return {
            'id': self.id,
            'key': self.key,
            'description': self.description,
            'title': self.title,
            'updatedAt': self.updatedAt,
            'createdAt': self.createdAt,
            'issueStatus': self.status.to_viewmodel() if self.status else None,
            'issueType': self.type.to_viewmodel() if self.type else None,
            'assignee': self.assignee.to_viewmodel() if self.assignee else None,
            'reporter': self.reporter.to_viewmodel() if self.reporter else None,
        }

    @classmethod
    def from_jira(cls, jira_obj):
        fields = jira_obj['fields']
        result = cls(
            jira_obj['id'],
            fields['summary'],
            fields['description'],
            fields['key'],
            fields['created'],
            fields['updated'],
        )
        result.assignee = User.from_jira(fields.get('assignee'))
        result.assignee = User.from_jira(fields.get('reporter'))
        return result
        # if (obj.fields.attachment) {
        #     for (let attachment of obj.fields.attachment) {
        #         result.attachments.push(AttachmentViewModel.createFromJira(attachment))
        #     }
        # }
        # result.issueStatus = StatusViewModel.createFromJira(obj.fields.status)
        # result.issueType = IssueTypeViewModel.createFromJira(obj.fields.issuetype, colorObj)
        # result.epic = EpicViewModel.createFromJira(obj.fields.epic)
        # result.tags = obj.fields.labels ? obj.fields.labels : []
        # const comments = obj.fields.comment ? obj.fields.comment.comments : []
        # for (let comment of comments) {
        #     result.comments.push(CommentViewModel.createFromJira(comment))
        # }
        # return result
